File: ios/AapSC/Services/SwimNotifications.swift
import Foundation
import UserNotifications

enum SwimNotifications {
    static let daysBeforeMonthEndToRemind = 5

    static func requestAuthorizationIfNeeded() async {
        let center = UNUserNotificationCenter.current()
        let settings = await center.notificationSettings()
        guard settings.authorizationStatus == .notDetermined else { return }
        _ = try? await center.requestAuthorization(options: [.alert, .sound, .badge])
    }

    static func refreshMonthlyGoalReminders(
        sessions: [SwimSession],
        profile: SwimProfile,
        monthlyChallengeRerolls: [String: MonthRerollEntry],
        t: TranslationService,
        now: Date = Date()
    ) async {
        await requestAuthorizationIfNeeded()
        let center = UNUserNotificationCenter.current()
        let settings = await center.notificationSettings()
        guard settings.authorizationStatus == .authorized
                || settings.authorizationStatus == .provisional else { return }

        let reminders = monthlyGoalReminders(
            sessions: sessions,
            profile: profile,
            monthlyChallengeRerolls: monthlyChallengeRerolls,
            t: t,
            now: now
        )

        let prefix = "monthly-goals-"
        let pending = await center.pendingNotificationRequests()
        for request in pending where request.identifier.hasPrefix(prefix) {
            center.removePendingNotificationRequests(withIdentifiers: [request.identifier])
        }

        guard let reminder = reminders.first else { return }

        let content = UNMutableNotificationContent()
        content.title = reminder.title
        content.body = reminder.body
        content.sound = .default

        var dateComponents = Calendar.current.dateComponents([.year, .month, .day, .hour], from: now)
        dateComponents.hour = max((dateComponents.hour ?? 9), 9)

        let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: false)
        let request = UNNotificationRequest(
            identifier: reminder.id,
            content: content,
            trigger: trigger
        )
        try? await center.add(request)
    }

    static func daysRemainingInMonth(_ date: Date = Date()) -> Int {
        let calendar = Calendar.current
        guard let range = calendar.range(of: .day, in: .month, for: date),
              let day = calendar.dateComponents([.day], from: date).day else {
            return 0
        }
        return max(0, range.count - day)
    }

    static func isNearMonthEnd(_ date: Date = Date()) -> Bool {
        daysRemainingInMonth(date) <= daysBeforeMonthEndToRemind
    }

    struct ReminderPayload: Equatable {
        var id: String
        var title: String
        var body: String
    }

    static func monthlyGoalReminders(
        sessions: [SwimSession],
        profile: SwimProfile,
        monthlyChallengeRerolls: [String: MonthRerollEntry],
        t: TranslationService,
        now: Date = Date()
    ) -> [ReminderPayload] {
        guard isNearMonthEnd(now) else { return [] }

        let monthKey = SwimMonthlyChallenges.getMonthKey(now)
        let mascotId = MascotUnlock.resolveMascotId(
            profile: profile,
            sessions: sessions,
            monthlyChallengeRerolls: monthlyChallengeRerolls
        )
        let intensity = MascotConstants.gameplay(mascotId).challengeIntensity
        let state = SwimMonthlyChallenges.evaluateMonthlyChallenges(
            sessions: sessions,
            monthKey: monthKey,
            rerolls: monthlyChallengeRerolls,
            intensity: intensity
        )
        let open = state.challenges.filter { !$0.completed }
        guard !open.isEmpty, state.completedCount < 3 else { return [] }

        let daysLeft = daysRemainingInMonth(now)
        let openSummary = open.prefix(2).map {
            SwimMonthlyChallengeFormatters.formatTarget(type: $0.type, target: $0.target, t: t)
        }.joined(separator: ", ")

        return [ReminderPayload(
            id: "monthly-goals-\(monthKey)",
            title: t.t("notifications.monthlyGoalsTitle"),
            body: t.t("notifications.monthlyGoalsBody", params: [
                "count": "\(open.count)",
                "days": "\(daysLeft)",
                "goals": openSummary,
            ])
        )]
    }
}

File: lib/swimNotifications.js
import { evaluateMonthlyChallenges, getMonthKey } from './swimMonthlyChallenges.js';
import { formatChallengeTarget } from './monthlyChallengeFormatters.js';
import { resolveMascotId, getMascotGameplay } from './mascotConstants.js';
import { daysSinceLastSession } from './swimLaunchSync.js';

export const DAYS_BEFORE_MONTH_END_TO_REMIND = 5;
export const UPLOAD_REMINDER_MIN_DAYS = 7;

const notificationStorageKey = 'AUDIT_SWIM_NOTIFICATIONS';

export const getDaysRemainingInMonth = (date = new Date()) => {
  const lastDay = new Date(date.getFullYear(), date.getMonth() + 1, 0);
  return Math.max(0, lastDay.getDate() - date.getDate());
};

export const isNearMonthEnd = (date = new Date(), daysBeforeEnd = DAYS_BEFORE_MONTH_END_TO_REMIND) =>
  getDaysRemainingInMonth(date) <= daysBeforeEnd;

const makeTr = (t) => (key, params = {}) => {
  let str = typeof t === 'function' ? t(key) : key;
  Object.entries(params).forEach(([k, v]) => {
    str = str.replaceAll(`{${k}}`, v != null && v !== '' ? String(v) : '—');
  });
  return str.replace(/\{[a-zA-Z]+\}/g, '—');
};

/** Build in-app / push reminder payloads for open monthly goals near month end. */
export const getMonthlyGoalReminders = (
  sessions,
  profile,
  t,
  { monthlyChallengeRerolls = {}, now = new Date() } = {}
) => {
  if (!isNearMonthEnd(now)) return [];

  const monthKey = getMonthKey(now);
  const mascotId = resolveMascotId(profile, { sessions, monthlyChallengeRerolls });
  const intensity = getMascotGameplay(mascotId).challengeIntensity;
  const state = evaluateMonthlyChallenges(sessions, monthKey, monthlyChallengeRerolls, intensity);
  const open = state.challenges.filter((c) => !c.completed);
  if (!open.length || state.completedCount >= 3) return [];

  const tr = makeTr(t);
  const daysLeft = getDaysRemainingInMonth(now);
  const openSummary = open
    .slice(0, 2)
    .map((c) => formatChallengeTarget(c.type, c.target, t))
    .join(', ');

  return [{
    id: `monthly-goals-${monthKey}`,
    type: 'monthlyGoals',
    title: tr('notifications.monthlyGoalsTitle'),
    body: tr('notifications.monthlyGoalsBody', {
      count: open.length,
      days: daysLeft,
      goals: openSummary,
    }),
    monthKey,
    openCount: open.length,
  }];
};

/** Remind web users to upload when no session has been logged recently. */
export const getUploadSyncReminder = (sessions, t, { now = new Date(), minDays = UPLOAD_REMINDER_MIN_DAYS } = {}) => {
  const days = daysSinceLastSession(sessions, now);
  if (days == null || days < minDays) return null;

  const tr = makeTr(t);
  return {
    id: 'upload-sync',
    type: 'uploadSync',
    title: tr('notifications.uploadSyncTitle'),
    body: tr('notifications.uploadSyncBody', { days }),
    daysSinceLast: days,
  };
};

export const collectLaunchReminders = (
  sessions,
  profile,
  t,
  options = {}
) => {
  const reminders = [...getMonthlyGoalReminders(sessions, profile, t, options)];
  const upload = getUploadSyncReminder(sessions, t, options);
  if (upload) reminders.push(upload);
  return reminders;
};

const readDismissed = () => {
  if (typeof window === 'undefined') return {};
  try {
    const raw = localStorage.getItem(notificationStorageKey);
    return raw ? JSON.parse(raw) : {};
  } catch {
    return {};
  }
};

export const dismissReminder = (id) => {
  if (typeof window === 'undefined') return;
  const dismissed = readDismissed();
  dismissed[id] = new Date().toISOString();
  localStorage.setItem(notificationStorageKey, JSON.stringify(dismissed));
};

export const filterUndismissedReminders = (reminders, { now = new Date() } = {}) => {
  const dismissed = readDismissed();
  return reminders.filter((r) => {
    const at = dismissed[r.id];
    if (!at) return true;
    const dismissedAt = new Date(at);
    if (Number.isNaN(dismissedAt.getTime())) return true;
    // Monthly goal reminders can reappear each day in the final week.
    if (r.type === 'monthlyGoals') {
      return now.toDateString() !== dismissedAt.toDateString();
    }
    // Upload reminder: re-show after 3 days.
    const days = Math.round((now - dismissedAt) / 86400000);
    return days >= 3;
  });
};

export const showBrowserNotification = async (reminder) => {
  if (typeof window === 'undefined' || !('Notification' in window)) return false;
  if (Notification.permission === 'denied') return false;
  if (Notification.permission === 'default') {
    const result = await Notification.requestPermission();
    if (result !== 'granted') return false;
  }
  try {
    new Notification(reminder.title, { body: reminder.body, tag: reminder.id });
    return true;
  } catch {
    return false;
  }
};
